enhancements/machine-api/control-plane.md

```diff
@@ -0,0 +1,339 @@
+- E2e testing on the machine API [e2e test suite](https://github.com/openshift/cluster-api-actuator-pkg/tree/master/pkg):
+	- Scale from 3 to 5.
+	- Scale from 5 to 3.
+	- Vertical scaling Rolling upgrade.
+	- Simulate 1 unhealthy node and remediation.
+- E2e testing on the origin e2e test suite.
+
+### Graduation Criteria
+
+**Note:** *Section not required until targeted at a release.*
+
+This controller will go GA in 4.N as long as:
+- All the testing above is in place.
+- There is considerable internal usage and manual disruptive QE testing feedback.
+- This is beta tested pre-release during a considerable period of time either internally or by an early adopter user through the dev preview builds.
+
+Otherwise this would go "Tech Preview" under a feature gate first until enough real use feedback is developed. The risk though is that there might be a very scarce use of feature gated features in the field as to develop the feedback we need.
+
+### Upgrade / Downgrade Strategy
+
+During the cluster upgrade for the targeted release the Machine API Operator (MAO) will let the CVO to instantiate the new CRD `controlPlane` and it will run the backing controller making this functionality opt-in for existing clusters. The user can create an instance of the new CRD if they choose to do so.
+
+New IPI clusters deployed after the targeted release will run the `controlPlane` instance deployed by the installer out of the box.
+
+### Version Skew Strategy
+
+## Implementation History
+- "https://github.com/kubernetes-sigs/cluster-api/blob/master/docs/proposals/20191017-kubeadm-based-control-plane.md"
+- "https://github.com/openshift/enhancements/blob/master/enhancements/etcd/cluster-etcd-operator.md"
+- "https://github.com/openshift/machine-config-operator/blob/master/docs/etcd-quorum-guard.md"
+
+## Drawbacks
+
+## Alternatives
+
+1. Use only existing resources for managing the Control Plane:
+
+This wouldn't be enough to satisfy the user stories in this proposal.
+
+Scenario 1: Scale out compute from 3 to 5. 2 etcd member healthy and 1 unhealthy out of 3.
+
+- With only machineSet + MHC:
+	- New machines start to come up.
+	- As soon as the etcd API is notified of a new member( total 4, healthy 2) the cluster
+	loses quorum until that new member starts and joins the cluster.
+	- There's no automation mechanism for Machines to spread evenly across failure domains.
+
+- With upper level controller:
+	- If (NOT all etcd members are healthy OR NOT all owned machines have a backed ready node) then controller short circuits here, log, update status and requeue. Else:
+	- Scale up one machine at a time. Go to 1.
+
+Scenario 2: 4 etcd member healthy and 1 unhealthy out of 5. Scale in compute from 5 to 3.
+
+- With only machineSet + MHC:
+	- Machines start to get deleted.
+	- etcd peer membership is not removed. 
+	- etcd guard blocks on drain before losing quorum. etcd remains degraded.
```

Comment:
since machine-api always drains, and etcd operator is the one that decides the disruption.
here:
pdb is 3 out of 5
one is already down.

machine-api is asked to go down from 3->5
one machine can actually go done.
the second machine cannot because of pdb making the drain fail.
the etcd is not degraded because it still have 3 up etcd members?